zkfacade/src/main/java/com/yahoo/vespa/curator/api/AbstractSingletonWorker.java

```diff
@@ -0,0 +1,95 @@
+package com.yahoo.vespa.curator.api;
+
+import com.yahoo.component.AbstractComponent;
+import com.yahoo.concurrent.UncheckedTimeoutException;
+import com.yahoo.vespa.curator.api.VespaCurator.SingletonWorker;
+import com.yahoo.yolean.UncheckedInterruptedException;
+
+import java.time.Duration;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Super-class for implementations of {@link SingletonWorker}.
+ * Users should call {@link VespaCurator#registerSingleton} at construction, and
+ * {@link VespaCurator#unregisterSingleton} at deconstruction.
+ * If this fails activation on registration, it will immediately unregister as well, before propagating the error.
+ * Consequently, registering this on construction will allow detecting a failed component generation, and instead
+ * retain the previous generation, provided enough verification is done in {@link #activate()}.
+ * The default ID to use with registration is the concrete class name, e.g., {@code my.example.Singleton}.
+ *
+ * @author jonmv
+ */
+public abstract class AbstractSingletonWorker implements SingletonWorker {
+
+    private final AtomicReference<VespaCurator> owner = new AtomicReference<>();
+
+    /**
+     * The singleton ID to use when registering this with a {@link VespaCurator}.
+     * At most one singleton worker with the given ID will be active, in the cluster, at any time.
+     * {@link VespaCurator#isActive(String)} may be polled to see whether this container is currently
+     * allowed to have an active singleton with the given ID.
+     */
+    public String id() { return getClass().getName(); }
+
+    /**
+     * <strong>Call this at the end of construction of the owner of this.</strong>
+     * If this activates the singleton, this happens synchronously, and any errors are propagated here.
+     * If this replaces an already active singleton, its deactivation is also called, prior to activation of this.
+     * If (de)activation is not complete within the given timeout, a timeout exception is thrown.
+     * If an error occurs (due to failed activation), unregistration is automatically attempted, with the same timeout.
+     */
+    public final void register(VespaCurator curator, Duration timeout) {
```

Comment:
Move this method to VespaCurator, to make a blocking variant of registerSingleton()?  There is no reason this method wouldn't just work on the more general SingletonWorker, as long as VespaCurator only allows the registering of each singleton instance once.

Reply:
The check for whether `owner` is currently unset becomes a kludge if it's moved.

Reply:
The only reason you need an owner is to detect double-registration, but VespaCurator.registerSingleton() already supports detecting double-registration?

Reply:
Only against the same curator.

Reply:
So it's to protect against registering an instance of SingletonWorker from being registered against two different VespaCurator?  How is that even possible?

Reply:
Anything's possible! But perhaps it's not needed. You'd need to do some really silly stuff to manage to pull that off, I suppose ...